content/blog/mcp-with-wasmcp.md

```diff
@@ -0,0 +1,387 @@
+```
+
+We compose our first and second tool components together by adding the paths to both tool component binaries in the `wasmcp compose server` arguments. Note that these local paths can be substituted for OCI registry artifacts. See `wasmcp compose server –help` for details.
+
+```
+$ wasmcp compose server ./my-first-tools/target/wasm32-wasip2/release/my-first-tools.wasm ./python-tools/python-tools.wasm -o polyglot.wasm
+
+```
+
+Run `polyglot.wasm` with `spin up`.
+```
+$ spin up -f polyglot.wasm
+
+Serving http://127.0.0.1:3000
+```
+
+Now our server has four tools: `add`, `subtract`, `reverse`, and `uppercase`! Two are implemented in Python, and two in Rust.
+
+### Wasmcp's architecture
+
+Server features like tools, resources, prompts, and completions, are implemented by individual WebAssembly components that export narrow [WIT](https://component-model.bytecodealliance.org/design/wit.html) interfaces mapped from the MCP spec's [schema types](https://modelcontextprotocol.io/specification/draft/schema).
+
+`wasmcp compose` plugs these feature components into framework components and composes them together behind a transport component as a complete middleware [chain of responsibility](https://en.wikipedia.org/wiki/Chain-of-responsibility_pattern) that implements an MCP server.
+
+Any of the wasmcp framework components, like the transport, can be swapped out for custom implementations during composition, enabling flexible server configurations.
+
+```
+Transport<Protocol>
+        ↓
+    Middleware₀
+        ↓
+    Middleware<Feature>₁
+        ↓
+    Middleware<Feature>₂
+        ↓
+       ...
+        ↓
+    Middlewareₙ
+        ↓
+    MethodNotFound
+```
+
+Each component:
+- Handles requests it understands (e.g., `tools/call`)
+- Delegates others downstream
+- Merges results (e.g., combining tool lists)
+
+This enables dynamic composition without complex configuration, all within a single Wasm binary. Think Unix pipes for MCP using Wasm components.
+
+This example only scratched the surface of what we can potentially do with `wasmcp`. To see some of the more advanced patterns like custom middleware components and session-enabled features, check out the [examples](https://github.com/wasmcp/wasmcp/tree/main/examples).
+
+## Publishing to OCI Registries
+
+We can use [wkg](https://github.com/bytecodealliance/wasm-pkg-tools) to publish our server to an [OCI](https://opencontainers.org/) registry like [GitHub Container Registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry), [Docker Hub](https://docs.docker.com/docker-hub/repos/manage/hub-images/oci-artifacts/), or [Amazon Elastic Container Registry](https://aws.amazon.com/ecr/).
+
+```
+$ wkg publish polyglot.wasm --package mygithub:basic-utils@0.1.0
+```
+
+Anyone with read access to this artifact can then download and run the server using `wkg`.
+
+```
+$ wkg get mygithub:basic-utils@0.1.0
```

Comment:
Do you need to fetch it with wkg before you can run it with spin?

Reply:
It would be great to support running these from oci ref. I moved the over to an issue on Spin  for this https://github.com/spinframework/spin/issues/3360. Thanks for documenting the workaround with `wkg`